src/Lean/Elab/Tactic/BVDecide/Frontend/Normalize/AC.lean

```diff
@@ -1,39 +1,408 @@
+      let P := mkApp2 (.const ``BEq.beq [uLvl]) ty inst
+      let out ← canonicalizeWithSharing P lhs rhs
+      return out
+  | Eq ty lhs rhs =>
+      trace[Meta.AC] "bv_ac_nf {checkEmoji} found `Eq`."
+      let uLvl ← getLevel ty
+      let P := mkApp (.const ``Eq [uLvl]) ty
+      let out ← canonicalizeWithSharing P lhs rhs
+      return out
+  | _ =>
+    return .continue
+
+/-! ## Tactic Boilerplate -/
+
+open Tactic
+
+def bvAcNfTarget (mvarId : MVarId)
+    (maxSteps : Nat := Lean.Meta.Simp.defaultMaxSteps) : MetaM MVarId := do
+  let simpCtx ← Simp.mkContext
+      (simpTheorems  := {})
+      (congrTheorems := (← getSimpCongrTheorems))
+      (config        := { Simp.neutralConfig with maxSteps})
+  let tgt ← instantiateMVars (← mvarId.getType)
+  let (res, _) ← Simp.main tgt simpCtx (methods := { post := bvAcNfpost })
+  applySimpResultToTarget mvarId tgt res
+
+
+def bvAcNfHypMeta (goal : MVarId) (fvarId : FVarId)
+    (maxSteps : Nat := Lean.Meta.Simp.defaultMaxSteps) : MetaM (Option MVarId) := do
+  goal.withContext do
+    let simpCtx ← Simp.mkContext
+      (simpTheorems  := {})
+      (congrTheorems := (← getSimpCongrTheorems))
+      (config        := { Simp.neutralConfig with maxSteps})
+    let tgt ← instantiateMVars (← fvarId.getType)
+    let (res, _) ← Simp.main tgt simpCtx (methods := { post := bvAcNfpost })
+    return (← applySimpResultToLocalDecl goal fvarId res false).map (·.snd)
 
 /--
-Normalize with respect to Associativity and Commutativity.
+Normalize the tactic target up to associativity and commutativity of bitvector
+multiplication, in a way that exposes common terms among both sides of an equality.
+
+This is similar to the `ac_nf` tactic, except that it is specialized to bitvector
+multiplication, and it differs in how the canonical form of the left-hand-side of
+an equality can depend on the right-hand-side, in particular, to expose shared terms.
+
+For example, `x₁ * (y₁ * z) = x₂ * (y₂ * z)` is normalized to
+`z * (x₁ * y₁) = z * (x₂ * y₂)`, pulling the shared variable `z` to the front on
+both sides.
 -/
-def acNormalizePass : Pass where
-  name := `ac_nf
+def bvAcNfTargetTactic : TacticM Unit := do
+  liftMetaTactic1 fun goal => bvAcNfTarget goal
+
+/--
+Normalize an hypothesis up to associativity and commutativity of bitvector
+multiplication, in a way that exposes common terms among both sides of an equality.
+
+This is similar to the `ac_nf` tactic, except that it is specialized to bitvector
+multiplication, and it differs in how the canonical form of the left-hand-side of
+an equality can depend on the right-hand-side, in particular, to expose shared terms.
+
+For example, `x₁ * (y₁ * z) = x₂ * (y₂ * z)` is normalized to
+`z * (x₁ * y₁) = z * (x₂ * y₂)`, pulling the shared variable `z` to the front on
+both sides.
+-/
+def bvAcNfHypTactic (fvarId : FVarId) : TacticM Unit :=
+  liftMetaTactic1 fun goal => bvAcNfHypMeta goal fvarId
+
+def bvAcNormalizePass (maxSteps : Nat) : Pass where
```

Comment:
You are still in the correct monad to get maxSteps within `run'`, there is no need for this parameter.